theano/tensor/nnet/abstract_conv2d.py

```diff
@@ -0,0 +1,741 @@
+
+    ## Update shape/height_width
+    def make_node(self, img, topgrad, shape):
+        if img.type.ndim != 4:
+            raise TypeError('img must be 4D tensor')
+        if topgrad.type.ndim != 4:
+            raise TypeError('topgrad must be 4D tensor')
+        if self.subsample != (1, 1) or self.border_mode == "half":
+            if shape is None:
+                raise ValueError('shape must be given if subsample != (1, 1)'
+                                 ' or border_mode == "half"')
+
+        shape = as_tensor_variable(shape)
+        broadcastable=[topgrad.broadcastable[1],
+                       img.broadcastable[1],
+                       False, False]
+        output = img.type.clone(broadcastable=broadcastable)()
+        return Apply(self, [img, topgrad, shape], [output])
+
+    def perform(self, node, inp, out_):
+        raise NotImplementedError('AbstractConv2d_gradWeight theano optimization failed')
+
+    def grad(self, inp, grads):
+        bottom, top = inp[:2]
+        weights, = grads
+        d_bottom = AbstractConv2d_gradInputs(self.imshp, self.kshp,
+                                             self.bsize,
+                                             self.border_mode,
+                                             self.subsample,
+                                             self.filters_flip)(weights, top, bottom.shape[-2:])
+        d_top = AbstractConv2d(self.imshp,
+                               self.kshp,
+                               self.bsize,
+                               self.border_mode,
+                               self.subsample,
+                               self.filters_flip)(bottom, weights)
+        d_height_width = (theano.gradient.DisconnectedType()(),)
+        return (d_bottom, d_top) + d_height_width
+
+    def connection_pattern(self, node):
+        return [[1], [1], [0]]  # no connection to height, width
+
+
+class AbstractConv2d_gradInputs(BaseAbstractConv2d):
+    """Gradient wrt. inputs for `AbstractConv2d`.
+
+    :note: You will not want to use this directly, but rely on
+           Theano's automatic differentiation or graph optimization to
+           use it as needed.
+
+    """
+
+    def __init__(self,
+                 imshp=None,
+                 kshp=None,
+                 bsize=None,
+                 border_mode="valid",
+                 subsample=(1, 1),
+                 filters_flip=True):
+        super(AbstractConv2d_gradInputs, self).__init__(imshp, kshp, bsize,
+                                                        border_mode, subsample, filters_flip)
+
+    ## Update shape/height_width
+    def make_node(self, kern, topgrad, shape):
+        if kern.type.ndim != 4:
+            raise TypeError('kern must be 4D tensor')
+        if topgrad.type.ndim != 4:
+            raise TypeError('topgrad must be 4D tensor')
+        if self.subsample != (1, 1) and shape is None:
+            raise ValueError('shape must be given if subsample != (1, 1)')
```

Comment:
No, you clearly don't. The very next line is `shape = as_tensor_variable(shape)`!